lib/blobsterix/storage/file_system.rb

```diff
@@ -7,93 +7,151 @@ class FileSystem < Storage
-        FileUtils.mkdir_p(@contents) unless Dir.exist?(@contents)
-        FileUtils.touch File.join(@contents,".keep")
+        @contents_path = path
+        @contents = nil
       end
 
       def bucket_exist(bucket="root")
-        Dir.entries(contents).include?(bucket) and File.directory?(File.join(contents,bucket))
+        begin
+          Dir.entries(contents).include?(bucket) and File.directory?(File.join(contents,bucket))
+        rescue => e
+          if e.is_a? ::Blosterix::BlobsterixStorageError
+            raise e
+          else
+            raise ::Blosterix::BlobsterixStorageError.new("Could not check for bucket")
+          end
+        end
       end
 
       def list(bucket="root", opts={})
-        if bucket =~ /root/
-          list_buckets
-        else
-          if bucket_exist(bucket)
-            b = Bucket.new(bucket, time_string_of(bucket))
-            b.marker = opts[:start_path] if opts[:start_path]
-            Blobsterix.wait_for(Proc.new {
-              collect_bucket_entries(bucket, b, opts)
-            })
-            b
+        begin
+          if bucket =~ /root/
+            list_buckets
           else
-            Nokogiri::XML::Builder.new do |xml|
-              xml.Error "no such bucket"
+            if bucket_exist(bucket)
+              b = Bucket.new(bucket, time_string_of(bucket))
+              b.marker = opts[:start_path] if opts[:start_path]
+              Blobsterix.wait_for(Proc.new {
+                collect_bucket_entries(bucket, b, opts)
+              })
+              b
+            else
+              Nokogiri::XML::Builder.new do |xml|
+                xml.Error "no such bucket"
+              end
             end
           end
+        rescue => e
+          if e.is_a? ::Blosterix::BlobsterixStorageError
+            raise e
+          else
+            raise ::Blosterix::BlobsterixStorageError.new("Could not list bucket(s)")
+          end
         end
       end
 
       def get(bucket, key)
-        if (not File.directory?(contents(bucket, key))) # and bucket_files(bucket).include?(key)
-          Blobsterix.storage_read(BlobAccess.new(:bucket => bucket, :id => key))
-          metaData(bucket, key)
-        else
-          Blobsterix.storage_read_fail(BlobAccess.new(:bucket => bucket, :id => key))
-          Blobsterix::Storage::BlobMetaData.new
+        begin
+          if (not File.directory?(contents(bucket, key))) # and bucket_files(bucket).include?(key)
+            Blobsterix.storage_read(BlobAccess.new(:bucket => bucket, :id => key))
+            metaData(bucket, key)
+          else
+            Blobsterix.storage_read_fail(BlobAccess.new(:bucket => bucket, :id => key))
+            Blobsterix::Storage::BlobMetaData.new
+          end
+        rescue => e
+          if e.is_a? ::Blosterix::BlobsterixStorageError
+            raise e
+          else
+            raise ::Blosterix::BlobsterixStorageError.new("Could not get bucket entry: #{@contents_path}")
```

Comment:
You might want to re-use the error handling and add a module here with a `handle_storage_error(e, msg)` method that does
```ruby
if e.is_a? ::Blosterix::BlobsterixStorageError
  raise e
else 
  raise ::Blosterix::BlobsterixStorageError.new(msg)
end
```
and re-use it for all cases.